man/common-lisp-controller.3
.\" 	-*- Mode: Nroff -*-
.TH "common-lisp-controller" "3" "March, 15, 2006"
.AT 3
.SH NAME
common-lisp-controller - Common Lisp system management tool

.SH DESCRIPTION

The
.I common-lisp-controller
system will manage the installed Common
Lisp
.I implementations
and 
.I libraries
on the system. The main function is to enable
every user to load any available 
.I library
and to store the compiled files in a user and
implementation specific cache directory.

On upgrades of either 
.I implementations
or 
.I libraries
these directories will be cleared so that no
old fasl files will remain.

There are also options to add libraries by default
to new images, see
.I /usr/share/doc/common-lisp-controller/README.Debian
, or to add user specific libraries to
the management system, see
.I clc-register-user-package.
 
.SH EXAMPLES

To load a clc library
.I :foo
on a clc enabled implementation simply do

.B (clc:clc-require :foo)

.SH FILES

.I /var/cache/common-lisp/
.IB <uid>
.I /
.IB <implementation-name>
.I /
.IB <library>
the cache directory for a library fir a given user 
with associated uid and a certain implementation.

.I /usr/share/common-lisp/systems/ 
the location of the libraries.

.SH BUGS

Bug reports should be sent to the Debian BTS, preferably with
reportbug.

.SH AUTHOR
Peter Van Eynde (pvaneynd@debian.org) 2001-2006

Kevin M. Rosenberg (kmr@debian.org) 2001-2004

Ren\['e] van Bevern 2005-2006

Desmond O. Chang (dochang@gmail.com) 2010

.SH "SEE ALSO"
.BR clc-register-user-package (1),
.BR clc-unregister-user-package (1)
.BR register-common-lisp-implementation (8)
.BR register-common-lisp-source (8)
.BR unregister-common-lisp-implementation|source (8)
.BR unregister-common-lisp-source (8)


debian/NEWS
common-lisp-controller (7.0) unstable; urgency=low

  This new version uses asdf version 1.502 or higher so it only needs to
  install files in /etc/common-lisp/source-registry.conf.d/ in order
  for asdf to find it.

  CLC also uses the standard 'place binaries there' feature of asdf to 
  put the binaries in /var/cache/common-lisp. The problem is that this
  feature is opaque to us, so we cannot remove specifc libraries 
  or files for one implementation. For now we will nuke all 
  the fasl files in the cache when we add or remove a library or an
  implementation.

  This also means that clc enabled packages are _always_ active in asdf.

  As a bonus there is now also clc-clbuild which is a small wrapper round
  clbuild enabling you to download clbuild enabled packages.

 -- Peter Van Eynde <pvaneynd@debian.org>  Sun, 31 Jan 2010 21:10:18 +0100

common-lisp-controller (6.21) unstable; urgency=low

  * From version 6.19 onwards clc does not longer add the 
    /usr/share/common-lisp/systems/ directory to asdf:*central-registry*
    
    The net effect of this is that just using "(asdf:oos 'asdf:load-op ...)" 
    will no longer work. The only way to load clc enabled packages now
    it to use "(clc:clc-require :<foo>)"
    
    You can also activate the clc packages using "clc:with-clc-active" so
    
    (clc:with-clc-active
      (asdf:oos 'asdf:load-op :<foo>))
    
    will work too.
    
    This change was done so that using for example clbuild in Debian becomes
    possible again. Otherwise the packaged systems will interfere with the
    clbuild or locally generated systems.  

 -- Peter Van Eynde <pvaneynd@debian.org>  Wed, 30 Dec 2009 22:44:19 +0100


common-lisp-controller.lisp
;;; -*- Mode: Lisp; Package: COMMON-LISP-CONTROLLER -*-
;;;
;;; Copyright (C) 2000,2010  Peter Van Eynde and Kevin M. Rosenberg
;;; Licensed under the LLGPL, see debian/copyright file


(in-package #:cl-user)

#+ecl ;; Hush ECL
(setf *load-verbose* nil
      *load-print* nil)

(defpackage #:common-lisp-controller
  (:use #:common-lisp)
  (:export #:init-common-lisp-controller
	   #:init-common-lisp-controller-v4
	   #:compile-common-lisp-controller-v5
	   #:init-common-lisp-controller-v5
	   #:clc-require
	   #:*clc-quiet*)
  (:nicknames #:clc))

(in-package #:common-lisp-controller)

(defvar *clc-quiet* nil
  "If true then clc prints no messages")

;; Some general utilities to make the
;; descriptions shorter

(defvar *fasl-root* nil "Root of implementation's directories of binary files")
(defvar *source-root* #p"/usr/share/common-lisp/source/"
	"Root of source directories")
(defvar *systems-root* #p"/usr/share/common-lisp/systems/"
        "Root of systems directory")
(defvar *image-preferences* #p"/etc/common-lisp/images/"
	"Directory where user can choose what systems shall be added to
Images per default")
(defvar *implementation-name* nil "The name of the implementation,
used to name the directory in /var/cache/common-lisp-controller")

(define-modify-macro appendf (&rest lists) append)

(defun init-common-lisp-controller-v5 (implementation-name)
  ;; register the systems root:
  (setf *implementation-name* implementation-name)
  
  (pushnew :common-lisp-controller *features*)
  (pushnew :clc-os-debian *features*))

(defun compile-common-lisp-controller-v5 (implementation-name)
  "Compiles the clc files. Returns a list of fasls
that should be loaded in the list to enable clc"
  (setf *implementation-name* implementation-name)

  (pushnew :common-lisp-controller *features*)
  (pushnew :clc-os-debian *features*)

  (let* ((fasl-root (merge-pathnames
		     (make-pathname
		      :directory
		      `(:relative "0" ,*implementation-name*))
		     #p"/var/cache/common-lisp-controller/")))

    (labels ((source-filename (package-name filename)
	       (let ((file (parse-namestring filename)))
		 (merge-pathnames
		  (make-pathname
		   :name (pathname-name file)
		   :type (pathname-type file)
		   :directory (list :relative package-name))
		  *source-root*)))
	     #+ecl
	     (system-fasl-filename (package-name filename)
	       ;; this is complex because ecl
	       ;; should produce system fasls,
	       ;; and they have .o extension
	       (merge-pathnames 
		(make-pathname :type "o")
		(fasl-filename package-name filename)))
	     (fasl-filename (package-name filename)
	       ;; this is complex because ecl
	       ;; should produce system fasls,
	       ;; and they have .o extension
	       (let* ((file (parse-namestring filename))
		      (output-path
		       (merge-pathnames
			(make-pathname :name (pathname-name file)
				       :type (pathname-type file)
				       :directory (list :relative package-name))
			fasl-root))
		      (compiled-file-pathname
		       (compile-file-pathname
			output-path)))
		 compiled-file-pathname))

	     (compile-and-load (package-name filename)
	       (let* ((file-path (source-filename package-name filename))
		      (compiled-file-pathname
		       (progn
			 ;; first make the target directory:
			 (ensure-directories-exist
			  (fasl-filename package-name filename))
			 ;; now compile it:
			 (compile-file file-path
				       :output-file (fasl-filename package-name filename)
				       :print nil
				       :verbose nil))))
		 ;; then load it:
		 (load compiled-file-pathname)
		 ;; return fasl filename
		 compiled-file-pathname
		 ;; now for ecl: make the system file
		 #+ecl 
		 (compile-file file-path
			       :output-file
			       (system-fasl-filename package-name filename)
			       :print nil
			       :verbose nil
			       ;; make 'linkable object files' 
			       :system-p t))))

      ;; then asdf:
      ;; For SBCL, take advantage of it's REQUIRE/contrib directories integration
      #+sbcl
      (when (boundp 'sb-ext::*module-provider-functions*)
	(pushnew :sbcl-hooks-require cl:*features*))
      
      ;; return a list
      (prog1
	  (nconc
	   (list

	    ;; first ourselves:
	    (compile-and-load  "common-lisp-controller"
			       "common-lisp-controller.lisp")
	    ;; asdf
	    (compile-and-load  "cl-asdf" "asdf.lisp")
	    
	    (compile-and-load  "cl-asdf" "wild-modules.lisp")
	    
	    ;; now patch it::
	    (compile-and-load "common-lisp-controller"
			      "post-sysdef-install.lisp"))

	   ;; so that it will neither recalculate it nor save it in our image
	   (let ((*fasl-root* fasl-root))
	     ;; "load-user-image-components" is in the above-loaded files.
	     (funcall (symbol-function
		       (find-symbol
			(symbol-name :load-user-image-components)
			:common-lisp-controller)))))

	#+sbcl
	(setq cl:*features* (delete :sbcl-hooks-require  cl:*features*))))))

(defun init-common-lisp-controller-v4 (implementation-name)
  "configures common-lisp-controller. IMPLEMENTATION-NAME
is the name of this implementation. Fasl's will be created in
/var/cache/common-lisp-controller/<userid>/<implementation>"
  (compile-common-lisp-controller-v5 implementation-name)
  ;; no need to load them as they are already loaded
  (init-common-lisp-controller-v5 implementation-name))

(defun init-common-lisp-controller (fasl-root
                                    &key
                                    (source-root "/usr/share/common-lisp/")
                                    (version 2))
  (declare (ignore source-root version))
  ;; vodoo: extract the name of the implementation
  ;; from the old fasl directory...
  (init-common-lisp-controller-v4
   (first
    (last
     (pathname-directory
      (parse-namestring
       fasl-root))))))
